server/sonar-server/src/main/java/org/sonar/server/qualityprofile/RuleActivator.java

```diff
@@ -423,118 +363,119 @@ private String validateParam(RuleParamDto ruleParam, @Nullable String value) {
+    checkArgument(builtInProfile.isBuiltIn(), "Rules profile with UUID %s is not built-in", builtInProfile.getKee());
 
-      } catch (BadRequestException e) {
-        // other exceptions stop the bulk activation
-        result.incrementFailed();
-        result.getErrors().addAll(e.errors());
-      }
+    RuleActivationContext.Builder builder = new RuleActivationContext.Builder();
+
+    // load rules
+    List<RuleDefinitionDto> rules = completeWithRules(dbSession, builder, ruleKeys);
+
+    // load profiles
+    List<QProfileDto> aliasedBuiltInProfiles = db.qualityProfileDao().selectQProfilesByRuleProfile(dbSession, builtInProfile);
+    List<QProfileDto> profiles = new ArrayList<>(aliasedBuiltInProfiles);
+    for (QProfileDto aliasProfile : aliasedBuiltInProfiles) {
+      // FIXME N+1 syndrome. To be optimized by replacing db column parent_kee by ancestor_keys.
+      profiles.addAll(db.qualityProfileDao().selectDescendants(dbSession, aliasProfile));
     }
-    activeRuleIndexer.commitAndIndex(dbSession, result.getChanges());
-    return result;
+    builder.setProfiles(profiles);
+    builder.setBaseProfile(builtInProfile);
+
+    // load active rules
+    Collection<String> ruleProfileUuids = Stream
+      .concat(Stream.of(builtInProfile.getKee()), profiles.stream().map(QProfileDto::getRulesProfileUuid))
+      .collect(MoreCollectors.toHashSet(profiles.size() + 1));
+    completeWithActiveRules(dbSession, builder, rules, ruleProfileUuids);
+
+    return builder.build();
   }
 
-  public BulkChangeResult bulkDeactivateAndCommit(DbSession dbSession, RuleQuery ruleQuery, QProfileDto profile) {
-    BulkChangeResult result = new BulkChangeResult();
-    Iterator<RuleKey> rules = ruleIndex.searchAll(ruleQuery);
-    while (rules.hasNext()) {
-      try {
-        RuleKey ruleKey = rules.next();
-        List<ActiveRuleChange> changes = deactivate(dbSession, profile, ruleKey);
-        result.addChanges(changes);
-        if (!changes.isEmpty()) {
-          result.incrementSucceeded();
-        }
-      } catch (BadRequestException e) {
-        // other exceptions stop the bulk activation
-        result.incrementFailed();
-        result.getErrors().addAll(e.errors());
-      }
+  public RuleActivationContext createContextForUserProfile(DbSession dbSession, QProfileDto profile, Collection<RuleKey> ruleKeys) {
+    checkArgument(!profile.isBuiltIn(), "Profile with UUID %s is built-in", profile.getKee());
+    RuleActivationContext.Builder builder = new RuleActivationContext.Builder();
+
+    // load rules
+    List<RuleDefinitionDto> rules = completeWithRules(dbSession, builder, ruleKeys);
+
+    // load descendant profiles
+    List<QProfileDto> profiles = new ArrayList<>(db.qualityProfileDao().selectDescendants(dbSession, profile));
+    profiles.add(profile);
+    if (profile.getParentKee() != null) {
+      profiles.add(db.qualityProfileDao().selectByUuid(dbSession, profile.getParentKee()));
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: "selectByUuid" is marked "javax.annotation.ParametersAreNonnullByDefault" but parameter 2 could be null. ([squid:S2637](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2637))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsb%2FSONAR-10052%2Fqprofiles&issues=AWE-qvP5mckxG3vIth75&open=AWE-qvP5mckxG3vIth75)

[ISSUE_KEY]: AWE-qvP5mckxG3vIth75